File: Python_Solutions/doubly_linked_list/doubly_linked_list.py
# Implement a doubly linked list data structure

class Node:
    def __init__(self, val):
        self.value = val
        self.next = None
        self.prev = None


class DoublyLinkedList:
    def __init__(self):
        self.head = None
        self.tail = None
        self.length = 0

    def push(self, val):
        newNode = Node(val)
        if self.length == 0:
            self.head = newNode
            self.tail = newNode
        else:
            self.tail.next = newNode
            newNode.prev = self.tail
            self.tail = newNode
        self.length += 1
        return True

    def pop(self):
        if self.length == 0:
            return None
        temp = self.tail
        if self.length == 1:
            self.head = None
            self.tail = None
        else:
            self.tail = self.tail.prev
            self.tail.next = None
            temp.prev = None
        self.length -= 1
        return temp

    def shift(self):
        if self.length == 0:
            return None
        temp = self.head
        if self.length == 1:
            self.head = 0
            self.tail = 0
        else:
            self.head = self.head.next
            self.head.prev = None
            temp.prev = None
        self.length -= 1
        return temp

    def unshift(self, val):
        newNode = Node(val)
        if self.length == 0:
            self.head = newNode
            self.tail = newNode
        else:
            self.head.prev = newNode
            newNode.next = self.head
            self.head = newNode
        self.length += 1
        return True

    def get(self, idx):
        if idx < 0 or idx >= self.length:
            return None
        if idx <= self.length // 2:
            current = self.head
            count = 0
            while count != idx:
                current = current.next
                count += 1
            return current
        else:
            current = self.tail
            count = self.length - 1
            while count != idx:
                current = current.prev
                count -= 1
            return current

    def set_val(self, idx, val):
        found = self.get(idx)
        if found is not None:
            found.value = val
            return True
        return False

    def insert(self, idx, val):
        if idx < 0 or idx > self.length:
            return False
        if idx == 0:
            return self.unshift(val)
        elif idx == self.length:
            return self.push(val)
        else:
            newNode = Node(val)
            beforeNode = self.get(idx-1)
            afterNode = beforeNode.next
            beforeNode.next = newNode
            newNode.prev = beforeNode
            newNode.next = afterNode
            afterNode.prev = newNode
            self.length += 1
            return True

    def remove(self, idx):
        if idx < 0 or idx >= self.length:
            return None
        if idx == 0:
            return self.shift()
        elif idx == self.length - 1:
            return self.pop()
        else:
            removedNode = self.get(idx)
            prevNode = removedNode.prev
            nextNode = removedNode.next
            prevNode.next = nextNode
            nextNode.prev = prevNode
            removedNode.prev = None
            removedNode.next = None
            self.length -= 1
            return removedNode


d = DoublyLinkedList()
d.push(1)
d.push(2)
print(d.pop().value)
print(d.pop().value)
print(d.pop())
d.push(1)
d.push(2)
d.push('Hello')
print(d.shift().value)
d.unshift('Hi there')
print(d.shift().value)

File: Js_Solutions/doubly_linked_list/doubly_linked_list.js
// Implement a doubly linked list data structure

class Node {
    constructor(val) {
        this.value = val;
        this.next = null;
        this.prev = null;
    }
}

class DoublyLinkedList {
    constructor() {
        this.head = null;
        this.tail = null;
        this.length = 0;
    }

    push(val) {
        const newNode = new Node(val);
        if (this.length === 0) {
            this.head = newNode;
            this.tail = newNode;
        } else {
            this.tail.next = newNode;
            newNode.prev = this.tail;
            this.tail = newNode;
        }
        this.length += 1;
        return true;
    }

    pop() {
        if (this.length === 0) {
            return undefined;
        }
        let temp = this.tail;
        if (this.length === 1) {
            this.head = null;
            this.tail = null;
        } else {
            this.tail = this.tail.prev;
            this.tail.next = null;
            temp.prev = null;
        }
        this.length--;
        return temp;
    }

    shift() {
        if (this.length === 0) {
            return undefined;
        }
        let temp = this.head;
        if (this.length === 1) {
            this.head = null;
            this.tail = null;
        } else {
            this.head = this.head.next;
            this.head.prev = null;
            temp.next = null;
        }
        this.length--;
        return temp;
    }

    unshift(val) {
        const newVal = new Node(val);
        if (this.head === 0) {
            this.head = newVal;
            this.tail = newVal;
        } else {
            this.head.prev = newVal;
            newVal.next = this.head;
            this.head = newVal;
        }
        this.length++;
        return true;
    }

    get(idx) {
        if (idx < 0 || idx >= this.length) {
            return null;
        }
        if (idx <= Math.floor(this.length / 2)) {
            let current = this.head;
            let count = 0;
            while (count != idx) {
                current = current.next;
                count++;
            }
            return current;
        } else {
            let current = this.tail;
            let count = this.length - 1;
            while (count != idx) {
                current = current.prev;
                count--;
            }
            return current;
        }
    }

    set(idx, val) {
        const found = this.get(idx);
        if (found !== null) {
            found.value = val;
            return true;
        }
        return false;
    }

    insert(idx, val) {
        if (idx < 0 || idx > this.length) {
            return false;
        }
        if (idx === 0) {
            return !!this.unshift(val);
        } else if (idx === this.length) {
            return !!this.push(val);
        } else {
            let newNode = new Node(val);
            let beforeNode = this.get(idx - 1);
            let afterNode = beforeNode.next;
            beforeNode.next = newNode;
            newNode.prev = beforeNode;
            newNode.next = afterNode;
            afterNode.prev = newNode;
        }
        this.length++;
        return true;
    }

    remove(idx) {
        if (idx < 0 || idx >= this.length) {
            return null;
        }
        if (idx === 0) {
            return this.shift();
        } else if (idx === this.length - 1) {
            return this.pop();
        } else {
            let removedNode = this.get(idx);
            let prevNode = removedNode.prev;
            let nextNode = removedNode.next;
            prevNode.next = nextNode;
            nextNode.prev = prevNode;
            this.length--;
            removedNode.prev = null;
            removedNode.next = null;
            return removedNode;
        }

    }
}

d = new DoublyLinkedList();
d.push(1)
d.push(2)
console.log(d);
// console.log(d.insert(1, 'hi'));
// console.log(d);
console.log(d.remove(1));
console.log(d);